pulsar-client-cpp/include/pulsar/MessageRoutingPolicy.h

```diff
@@ -33,7 +34,10 @@ class MessageRoutingPolicy {
  public:
     virtual ~MessageRoutingPolicy() {}
 
+    /** @deprecated */
     virtual int getPartition(const Message& msg) = 0;
+
+    virtual int getPartition(const Message& msg, const TopicMetadata& topicMetadata) = 0;
```

Comment:
I think this will break the compatibility for existing apps that have the custom router. 

For example, if I had : 
```cpp
class MyRouter : public MessageRoutingPolicy {
public:
  int getPartition(const Message& msg) { 
     // My custom routing logic
  }
};
```
This  will get a link error at runtime with the new Pulsar library.

I think the new method should have a default implementation that uses the other one, eg: 

```cpp
virtual int getPartition(const Message& msg, const TopicMetadata& topicMetadata)  {
  return getPartition(msg);
}
```

At the same time, I'm not sure we should leave the old `int getPartition(const Message& msg)` with no implementation, because that would force to provide both implementations to the user. 

Maybe we could default it to an invalid partition, like : 

```cpp
/** @deprecated 
   Use int getPartition(const Message& msg, const TopicMetadata& topicMetadata)
 */
virtual int getPartition(const Message& msg) {
   return -1;
}